From c4abb0d017d14d1000ade2b5628e742d700f16a3 Mon Sep 17 00:00:00 2001
From: Niranjan Pendharkar <npendhar@codeaurora.org>
Date: Mon, 24 Jul 2017 09:54:07 -0700
Subject: [PATCH 3/3] netd: recognize NDK type in ndc network api

ndc network command now accepts strings describing net_handle_t.
For e.g. ndc network create handle42966108894

Test: manual
Bug: 36682246
CRs-fixed: 2070022
Change-Id: I44857af3d82ea408ae429f9fe16fcf5c876851ae
(cherry picked from commit 4c18bd9164104e0a1a81ea505d569937edb30afa)
---
 server/CommandListener.cpp |  6 ++++++
 server/NetworkController.h | 31 +++++++++++++++++++++++++++++++
 2 files changed, 37 insertions(+)

diff --git a/server/CommandListener.cpp b/server/CommandListener.cpp
index b5aff11..77acf4f 100644
--- a/server/CommandListener.cpp
+++ b/server/CommandListener.cpp
@@ -69,6 +69,12 @@ unsigned stringToNetId(const char* arg) {
             return NetworkController::MIN_OEM_ID + n;
         }
         return NETID_UNSET;
+    } else if (!strncmp(arg, "handle", 6)) {
+        unsigned n = netHandleToNetId((net_handle_t)strtoull(arg + 6, NULL, 10));
+        if (NetworkController::MIN_OEM_ID <= n && n <= NetworkController::MAX_OEM_ID) {
+            return n;
+        }
+        return NETID_UNSET;
     }
     // strtoul() returns 0 on errors, which is fine because 0 is an invalid netId.
     return strtoul(arg, NULL, 0);
diff --git a/server/NetworkController.h b/server/NetworkController.h
index 48d5799..ceb538e 100644
--- a/server/NetworkController.h
+++ b/server/NetworkController.h
@@ -17,6 +17,7 @@
 #ifndef NETD_SERVER_NETWORK_CONTROLLER_H
 #define NETD_SERVER_NETWORK_CONTROLLER_H
 
+#include <android/multinetwork.h>
 #include "NetdConstants.h"
 #include "Permission.h"
 
@@ -33,6 +34,36 @@ struct android_net_context;
 namespace android {
 namespace net {
 
+// Utility to convert from netId to net_handle_t. Doing this here as opposed to exporting
+// from net.c as it may have NDK implications. Besides no conversion available in net.c for
+// obtaining handle given netId.
+// TODO: Use getnetidfromhandle() in net.c.
+static inline unsigned netHandleToNetId(net_handle_t fromNetHandle) {
+    const uint32_t k32BitMask = 0xffffffff;
+    // This value MUST be kept in sync with the corresponding value in
+    // the android.net.Network#getNetworkHandle() implementation.
+    const uint32_t kHandleMagic = 0xfacade;
+
+    // Check for minimum acceptable version of the API in the low bits.
+    if (fromNetHandle != NETWORK_UNSPECIFIED &&
+        (fromNetHandle & k32BitMask) != kHandleMagic) {
+        return 0;
+    }
+
+    return ((fromNetHandle >> (CHAR_BIT * sizeof(k32BitMask))) & k32BitMask);
+}
+
+// Utility to convert from nethandle to netid, keep in sync with getNetworkHandle
+// in Network.java.
+static inline net_handle_t netIdToNetHandle(unsigned fromNetId) {
+    const net_handle_t HANDLE_MAGIC = 0xfacade;
+
+    if (!fromNetId) {
+        return NETWORK_UNSPECIFIED;
+    }
+    return (((net_handle_t)fromNetId << 32) | HANDLE_MAGIC);
+}
+
 class DumpWriter;
 class Network;
 class UidRanges;
-- 
2.1.2

